src/lib/db.ts

```diff
@@ -1,19 +1,44 @@
-import { types, Client, ClientConfig, Pool } from 'pg'
+import { types, Pool, PoolConfig } from 'pg'
 import { PostgresMetaResult } from './types'
 
 types.setTypeParser(20, parseInt)
 
-export const init = (
-  config: ClientConfig,
-  { pooled = true } = {}
-): ((sql: string) => Promise<PostgresMetaResult<any>>) => {
-  const client = pooled ? new Pool(config) : new Client(config)
-  return async (sql: string) => {
-    try {
-      const { rows } = await client.query(sql)
-      return { data: rows, error: null }
-    } catch (e) {
-      return { data: null, error: { message: e.message } }
-    }
+export const init: (
+  config: PoolConfig
+) => {
+  query: (sql: string) => Promise<PostgresMetaResult<any>>
+  end: () => Promise<void>
+} = (config) => {
+  // XXX: Race condition could happen here: one async task may be doing
+  // `pool.end()` which invalidates the pool and subsequently all existing
+  // handles to `query`. Normally you might only deal with one DB so you don't
+  // need to call `pool.end()`, but since the server needs this, we make a
+  // compromise: if we run `query` after `pool.end()` is called (i.e. pool is
+  // `null`), we temporarily create a pool and close is right after.
+  let pool: Pool | null = new Pool(config)
```

Comment:
Interesting. We haven't seen this at all yet as far as I can tell. Perhaps there is somethign smart going on where there is a pool created per-config? 

So that means the race condition would only really happen if 2 different users are using the same DB at the same time?

Reply:
This doesn't really apply without the cache (since we're creating a new pool on each request). But with the cache it's possible that one async function gets a handle to the pool, yield execution through await, then another async function closes the pool through `pool.end()`.

Should be a rare case though as `pool.end()` is only called when the cache disposes the pool, i.e. the pool hasn't been used for a while right when this is happening.